Remove Python dependency (toPyTuple/fromPyTuple, jitCompilationUnit) in rref_impl.h/cpp

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32753 Remove Python dependency (toPyTuple/fromPyTuple, jitCompilationUnit) in rref_impl.h/cpp**

Functions to be bound as an Aten operator could not have Python dependency.

This is to refactor and remove Python dependency.

Differential Revision: [D5741675](https://our.internmc.facebook.com/intern/diff/D5741675/)

xush6528 committed:
[1/x] Remove Python dependency in RRef impl

Functions to be bound as an Aten operator could not have Python dependency.

This is to refactor and remove Python dependency.

Differential Revision: [D5741675](https://our.internmc.facebook.com/intern/diff/D5741675/)

[ghstack-poisoned]

xush6528 committed:
Update on "[1/x] Remove Python dependency in RRef impl"


Functions to be bound as an Aten operator could not have Python dependency.

This is to refactor and remove Python dependency.

Differential Revision: [D5741675](https://our.internmc.facebook.com/intern/diff/D5741675/)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 19d72a64:


**None of the build failures appear to be your fault**.

* **1/1** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=d119de8abde8ba436e429d66764fce0624e0e9e6&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base d119de8a since Jan 29
  <details>
  <summary>
  Please rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=19d72a64ef5411cee9b75df76a35fe5ad05374e9).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably **caused by upstream breakages**:
* [pytorch_linux_xenial_py3_6_gcc5_4_test](https://dr.pytorch.org/build-details.html?build_id=92735149) from Jan 29



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

xush6528 committed:
Update on "[1/x] Remove Python dependency in RRef impl"


Functions to be bound as an Aten operator could not have Python dependency.

This is to refactor and remove Python dependency.

Differential Revision: [D5741675](https://our.internmc.facebook.com/intern/diff/D5741675/)

[ghstack-poisoned]

mrshenli reviewed (commented):
LGTM and thanks for fixing var names. Will stamp when we get rid of irrelevant test failures.

zhaojuanmao reviewed (commented):
lgtm, since rref_impl.cpp has no python dependency any more, we can move this file from libtorch_python_sources to libtorch_sources in build_variables.py and CMakeLists.txt?

xush6528 committed:
Update on "Remove Python dependency in rref_impl.h/cpp"


Functions to be bound as an Aten operator could not have Python dependency.

This is to refactor and remove Python dependency.

Differential Revision: [D5741675](https://our.internmc.facebook.com/intern/diff/D5741675/)

[ghstack-poisoned]

mrshenli reviewed (approved):
Stamp to unblock.

xush6528 committed:
Update on "Remove Python dependency (toPyTuple/fromPyTuple, jitCompilationUnit) in rref_impl.h/cpp"


Functions to be bound as an Aten operator could not have Python dependency.

This is to refactor and remove Python dependency.

Differential Revision: [D5741675](https://our.internmc.facebook.com/intern/diff/D5741675/)

[ghstack-poisoned]

xush6528 committed:
Update on "Remove Python dependency (toPyTuple/fromPyTuple, jitCompilationUnit) in rref_impl.h/cpp"


Functions to be bound as an Aten operator could not have Python dependency.

This is to refactor and remove Python dependency.

Differential Revision: [D5741675](https://our.internmc.facebook.com/intern/diff/D5741675/)

[ghstack-poisoned]

xush6528 committed:
Update on "Remove Python dependency (toPyTuple/fromPyTuple, jitCompilationUnit) in rref_impl.h/cpp"


Functions to be bound as an Aten operator could not have Python dependency.

This is to refactor and remove Python dependency.

Differential Revision: [D5741675](https://our.internmc.facebook.com/intern/diff/D5741675/)

[ghstack-poisoned]

xush6528 commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -38,7 +90,17 @@ py::object PyRRef::toHere() {
+          PyObjectType::get());
+    } else {
+      value = std::move(rawValues).front();
+    }
```
This is a new change required to make rref_impl.cpp Python free.
Essentially, we are deferring type-specific handling to Python wrapper layer.

@mrshenli @zhaojuanmao

mrshenli replied:
This LGTM. If we need to expose C++ APIs in the future, we can come back to revisit the `toHere` API. This change should be OK as both `py_rref` and `rref_impl` are purely internal stuff.

mrshenli commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -8,6 +8,58 @@
+
+namespace {
+
+namespace {
```
Do we need this nested anonymous namespace here?

xush6528 replied:
No, I think, one anonymous namespace should have confined names into the same file.

mrshenli reviewed (commented):
Do we need to add `TORCH_API` to the classes/functions moved to libtorch?

mrshenli commented:
Yep, test failure shows we need `TORCH_API`:

```
Jan 30 22:52:49 ImportError: /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so: undefined symbol: _ZN5torch11distributed3rpc8callback18confirmPendingUserERKNS1_7MessageERKN3c108optionalINS_5utils11FutureErrorEEE
```

mrshenli reviewed (changes requested):
block until `TORCH_API` is added

xush6528 committed:
Update on "Remove Python dependency (toPyTuple/fromPyTuple, jitCompilationUnit) in rref_impl.h/cpp"


Functions to be bound as an Aten operator could not have Python dependency.

This is to refactor and remove Python dependency.

Differential Revision: [D5741675](https://our.internmc.facebook.com/intern/diff/D5741675/)

[ghstack-poisoned]

mrshenli reviewed (approved):
Stamp to unblock. Let's wait for all tests before merging.

xush6528 commented:
1 unrelated failure,

pytorch_linux_backward_compatibility_check_test
```
Jan 30 23:38:29 Broken ops: [
Jan 30 23:38:29 	aten::join(str self, str[] values) -> (str)
Jan 30 23:38:29 ]
```
